camel/models/base_model.py

```diff
@@ -30,25 +32,75 @@ def __init__(self, model_type: ModelType,
             model_config_dict (Dict[str, Any]): A config dictionary.
         """
         self.model_type = model_type
+
         self.model_config_dict = model_config_dict
+        self.check_model_config()
+
+        self._token_counter: Optional[BaseTokenCounter] = None
+
+    @property
+    def token_counter(self) -> BaseTokenCounter:
+        r"""Initialize the token counter for the model backend.
+
+        Returns:
+            BaseTokenCounter: The token counter following the model's
+                tokenization style.
+        """
+        if not self._token_counter:
+            self._token_counter = self.initialize_token_counter()
+        return self._token_counter
 
     @abstractmethod
     def run(self, messages: List[Dict]) -> Dict[str, Any]:
         r"""Runs the query to the backend model.
 
         Args:
-            messages (List[Dict]): message list with the chat history
+            messages (List[Dict]): Message list with the chat history
                 in OpenAI API format.
 
         Raises:
-            RuntimeError: if the return value from OpenAI API
-            is not a dict that is expected.
+            RuntimeError: If the return value from OpenAI API
+                is not a dict that is expected.
 
         Returns:
             Dict[str, Any]: All backends must return a dict in OpenAI format.
         """
         pass
 
+    @abstractmethod
+    def check_model_config(self):
+        r"""Check whether the input model configuration contains unexpected
+        arguments
+
+        Raises:
+            ValueError: If the model configuration dictionary contains any
+                unexpected argument for this model class.
+        """
+        pass
+
+    @abstractmethod
+    def initialize_token_counter(self) -> BaseTokenCounter:
```

Comment:
This method looks weird in the abstract class for a model backend. Can we remove it?

Reply:
emm...🤔Why is this weird? With this method I can just define the `token_counter` property in `BaseModel` without further modification in child classes. The child classes can focus on implementing how they initialize the token counter.

Reply:
BaseModelBackend is an abstract class that specifies a so called contract between a user and an implementation. Some helper methods should not be there in the contract and are implementation-specific. Property token_counter is fine and is a good new interface method. But it must be it. Without strings attached like initialize_blah_blah.